src/decimal.rs

```diff
@@ -2217,239 +2572,72 @@ impl<'a> AddAssign<&'a Decimal> for Decimal {
-            remainder = div_by_u32(&mut product, 10u32);
-            final_scale -= 1;
+    #[inline(always)]
+    fn sub(self, other: &Decimal) -> Decimal {
+        match self.checked_sub(*other) {
+            Some(diff) => diff,
+            None => panic!("Subtraction overflowed"),
         }
+    }
+}
 
-        // Round up the carry if we need to
-        if remainder >= 5 {
-            for part in product.iter_mut() {
-                if remainder == 0 {
-                    break;
-                }
-                let digit: u64 = u64::from(*part) + 1;
-                remainder = if digit > 0xFFFF_FFFF { 1 } else { 0 };
-                *part = (digit & 0xFFFF_FFFF) as u32;
-            }
-        }
+impl SubAssign for Decimal {
+    fn sub_assign(&mut self, other: Decimal) {
+        let result = self.sub(other);
+        self.lo = result.lo;
+        self.mid = result.mid;
+        self.hi = result.hi;
+        self.flags = result.flags;
+    }
+}
 
-        // If we're still above max precision then we'll try again to
-        // reduce precision - we may be dealing with a limit of "0"
-        if final_scale > MAX_PRECISION {
-            // We're in an underflow situation
-            // The easiest way to remove precision is to divide off the result
-            while final_scale > MAX_PRECISION && !is_all_zero(&product) {
-                div_by_u32(&mut product, 10);
-                final_scale -= 1;
-            }
-            // If we're still at limit then we can't represent any
-            // siginificant decimal digits and will return an integer only
-            // Can also be invoked while representing 0.
-            if final_scale > MAX_PRECISION {
-                final_scale = 0;
-            }
-        } else if !(product[3] == 0 && product[4] == 0 && product[5] == 0) {
-            // We're in an overflow situation - we're within our precision bounds
-            // but still have bits in overflow
-            panic!("Multiplication overflowed");
-        }
+impl<'a> SubAssign<&'a Decimal> for Decimal {
+    fn sub_assign(&mut self, other: &'a Decimal) {
+        Decimal::sub_assign(self, *other)
+    }
+}
 
-        Decimal {
-            lo: product[0],
-            mid: product[1],
-            hi: product[2],
-            flags: flags(negative, final_scale),
+impl<'a> SubAssign<Decimal> for &'a mut Decimal {
+    fn sub_assign(&mut self, other: Decimal) {
+        Decimal::sub_assign(*self, other)
+    }
+}
+
+impl<'a> SubAssign<&'a Decimal> for &'a mut Decimal {
+    fn sub_assign(&mut self, other: &'a Decimal) {
+        Decimal::sub_assign(*self, *other)
+    }
+}
+
+forward_all_binop!(impl Mul for Decimal, mul);
+
+impl<'a, 'b> Mul<&'b Decimal> for &'a Decimal {
+    type Output = Decimal;
+
+    #[inline]
+    fn mul(self, other: &Decimal) -> Decimal {
+        match self.checked_mul(*other) {
+            Some(prod) => prod,
+            None => panic!("Multiplication overflowed"),
```

Comment:
I like the idea of being more explicit with the errors (similar to `div` below) but that being said, that is more of an implementation issue that we can evolve towards. Accordingly, as we evolve we could probably keep the message attached to the result too (e.g. `imp Display`).